### PR TITLE
building Java csvSink

### PR DESCRIPTION
**Pull request by Nauna**
- This work allows the csvSink to be built and run on the server.
- Allows each property to have a javaToCSV/javaToCSVLabel  which is a string that will output specified formatting for the csv outputter(csvSink).
- clean up a few toCSV legacy formatting and overrides

goes hand in hand with this pr: https://github.com/nanoPayinc/NANOPAY/pull/7306

### CODE REVIEW

**Comment by Nauna** [comment]
Current State of the world

image

File: src/foam/dao/CSVSink.js [hunk 1]
```diff
@@ -44,13 +49,33 @@ foam.CLASS({
...
+      args: [
+        { name: 'value' }
+      ],
+      code: function(value) {
+        if ( ! this.isNewLine ) this.csv += ',';
+        this.isNewLine = false;
+        this.output_(value);
+      },
+      javaCode: `
+        StringBuilder sb = new StringBuilder();
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: This isn't the right use of a StringBuilder. You'll want to create a property that you store the StringBuilder in and this method will be something like:
Under properties:
```
{
  name: 'sb',
  flags: ['java'],
  javaType: 'StringBuilder',
  javaFactory: 'return new StringBuilder();'
}
```
and this method turns into:
```
        if ( ! getIsNewLine() ) getSb().append(",");
        setIsNewLine(false);
        output_(value);
```
and then you can implement the sink's `eof` method in java:
```
{
  name: 'eof'
  javaCode: 'setCsv(getSb().toString());'
}
```
And in the other methods, don't create new string builders and just re-use getSb() to append.

File: src/foam/dao/CSVSink.js [hunk 2]
```diff
@@ -76,34 +101,97 @@ foam.CLASS({
...
           }, function(value) {
             this.output_(value.toString());
-          })
+        }),
+      javaCode: `
+        StringBuilder sb = new StringBuilder();
+        sb.append(getCsv());
+        String s = value.toString();
+
+        if (s.indexOf("\\"") != -1 || s.indexOf(",") != -1) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: You might want to use logic similar to the code block above to keep js and java behavior more consistent with each other. You can't use an mmethod but you can use if/else similar to the java json outputter: https://github.com/foam-framework/foam2/blob/master/src/foam/lib/json/Outputter.java#L213

File: src/foam/dao/CSVSink.js [hunk 3]
```diff
@@ -76,34 +101,97 @@ foam.CLASS({
...
     },
+    {
+      name: 'put',
+      code: function(obj) {
+        if ( ! this.of ) this.of = obj.cls_;
 
-    function put(obj) {
-      if ( ! this.of ) this.of = obj.cls_;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((element) => {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Did you break this? Aren't props a StringArray now?

File: src/foam/java/PropertyInfo.js [hunk 4]
```diff
@@ -285,6 +287,20 @@ foam.CLASS({
               { name: 'obj', type: 'foam.core.FObject' }
             ],
             body: this.validateObj
+          },
+          {
+            name: 'toCSV',
+            visibility: 'public',
+            type: 'void',
+            args: [{ name: 'x', type: 'foam.core.X' }, { name: 'obj', type: 'Object' }, { name: 'outputter', type: 'foam.dao.CSVSink' }, { name: 'prop', type: 'Object' }],
```
Comment (mcarcaso) [inline src/foam/java/PropertyInfo.js]: Use multiple lines for this so it's easier to read.

File: src/foam/java/PropertyInfo.js [hunk 5]
```diff
@@ -285,6 +287,20 @@ foam.CLASS({
...
+            body: this.toCSV
+          },
+          {
+            name: 'toCSVLabel',
+            visibility: 'public',
+            type: 'void',
+            args: [{ name: 'outputter', type: 'foam.dao.CSVSink' }, { name: 'prop', type: 'Object' }],
```
Comment (mcarcaso) [inline src/foam/java/PropertyInfo.js]: Here too

File: src/foam/nanos/controller/ApplicationController.js [hunk 6]
```diff
@@ -60,6 +60,7 @@ foam.CLASS({
     'appConfig',
     'as ctrl',
     'currentMenu',
+    'filteredTableColumns',
```
Comment (mcarcaso) [inline src/foam/nanos/controller/ApplicationController.js]: You don't want this on the ApplicationController. This ties the ApplicationController to one set of filteredTableColumns but there's no reason we can't render multiple table views at once. I think troy's suggestion was to put this on `foam.comics.DAOController`

**Review by mcarcaso — CHANGES_REQUESTED** [review]

File: src/foam/dao/CSVSink.js (hunk 1 above)
Comment (Nauna) [reply src/foam/dao/CSVSink.js]: nice!

File: src/foam/dao/CSVSink.js (hunk 3 above)
Comment (Nauna) [reply src/foam/dao/CSVSink.js]: :( ya i did. Fixed with next push

File: src/foam/comics/DAOController.js [hunk 7]
```diff
@@ -253,16 +261,16 @@ foam.CLASS({
...
-        link.click();
-      });
+      this.csvDriver.exportDAO(this.__context__, data, this.filteredTableColumns)
+        .then(function(result) {
+          result = 'data:text/csv;charset=utf-8,' + result;
+          var encodedUri = encodeURI(result);
+          var link = document.createElement('a');
+          link.setAttribute('href', encodedUri);
+          link.setAttribute('download', 'data.csv');
+          document.body.appendChild(link);
```
Comment (adamvy) [inline src/foam/comics/DAOController.js:245]: We should be cleaning up this link after the fact as well

File: src/foam/dao/CSVSink.js [hunk 8]
```diff
@@ -76,34 +106,99 @@ foam.CLASS({
...
     },
-
-    function newLine_() {
-      this.csv += '\n';
-      this.isNewLine = true;
+    {
+      name: 'newLine_',
+      code: function() {
+        this.csv += '\n';
+        this.isNewLine = true;
```
Comment (adamvy) [inline src/foam/dao/CSVSink.js]: I personally would have called this isFirst, but that's not major

Comment (adamvy) [reply src/foam/dao/CSVSink.js]: Or maybe isFirstColumn

File: src/foam/dao/CSVSink.js [hunk 9]
```diff
@@ -76,34 +106,99 @@ foam.CLASS({
...
+      name: 'eof',
+      javaCode: 'setCsv(getSb().toString());'
+    },
+    {
+      name: 'put',
+      code: function(obj) {
+        if ( ! this.of ) this.of = obj.cls_;
+        var element = undefined;
+        if ( ! this.isHeadersOutput ) {
+          this.props.forEach((name) => {
```
Comment (adamvy) [inline src/foam/dao/CSVSink.js]: I would have done isFirstRow and put the header outputting into its own function.

`if ( this.isFirstRow ) this.outputHeaders`

Not a big deal

File: src/foam/lib/csv/Outputter.java [hunk 10]
```diff
@@ -176,7 +176,7 @@ protected void outputFObject(FObject obj) {
 
     while ( i.hasNext() ) {
       PropertyInfo prop = (PropertyInfo) i.next();
-      prop.toCSV(this, prop.f(obj));
+      output(prop.f(obj));
```
Comment (adamvy) [inline src/foam/lib/csv/Outputter.java]: Why this change?  shouldn't the property's toCSV still be respected if it wants to control how it outputs itself to CSV?

File: src/foam/u2/view/UnstyledTableView.js [hunk 11]
```diff
@@ -33,7 +33,8 @@ foam.CLASS({
     'ctrl',
     'dblclick?',
     'editRecord?',
-    'selection? as importSelection'
+    'selection? as importSelection',
+    'filteredTableColumns'
```
Comment (adamvy) [inline src/foam/u2/view/UnstyledTableView.js]: Make this an optional import

File: src/foam/u2/view/UnstyledTableView.js [hunk 12]
```diff
@@ -53,21 +54,28 @@ foam.CLASS({
...
-            of.getAxiomByName(p) :
-            p;
-
-           if ( ! c ) {
-             console.error('Unknown table column: ', p);
-           }
-
-          return c;
-        }).filter(function(c) { return c; });
+        this.filteredTableColumns = [];
```
Comment (adamvy) [inline src/foam/u2/view/UnstyledTableView.js]: Build the array first as a local variable, and then set the imported value afterwards.  This is for two reasons.  1) Anybody listening on updates to filteredTableColumns would first get a property change even as its set to an empty array. and then 2) property change events don't fire when an array's contents change.

Comment (adamvy) [reply src/foam/u2/view/UnstyledTableView.js]: Actually you'd be better off implementing this as a reaction.

```
foam.CLASS({
  name: 'UnstyledTableView',
...
  imports: [ 'filteredTableColumns?' ],
  reactions: [ '', 'propertyChange.columns_', 'onColumnsUpdate' ],
  listeners: [
    function onColumnsUpdate() {
      // First check if the optional import was provided.
      if ( this.filteredTableColumns$ ) {
        this.filteredTableColumns$ = this.columns_.map(p => p.name);
      }
    }
  ]
});
```

Then you don't need to modify the expression at all, and if somebody explicitly sets columns_ then filteredTableColumns will still be updated appropriately.

File: src/foam/lib/csv/Outputter.java (hunk 10 above)
Comment (mcarcaso) [reply src/foam/lib/csv/Outputter.java]: The behaviour of toCSV has changed and isn't really compatible with this Outputter anymore so I'm guessing the safest thing to do was this. If I'm reading this correctly, this roughly preserves the previous behaviour with the exception of Enums. IMO, we should be aiming to nuke this outputter completely as it's replaced by the CSVSink.

Comment (adamvy) [reply src/foam/lib/csv/Outputter.java]: I see that toCSV now takes a CSVSink?  That doesn't sound right to me.  CSVSink should just internally use a CSVOutputter.  As it stands we're setting outselves up for confusion and possible bugs as people write javaToCSV for properties but then that doesn't actually get called when the object is output the CSV.

Comment (Nauna) [reply src/foam/lib/csv/Outputter.java]: So short of re-writing this outputter to function like the new outputter, this was the quick fix.

This I feel should be removed. But need to confirm that no one is using it.

The toCSV that this is expecting is not the same as what we have developed here.

And ultimately all this toCSV that was doing was 'output(prop.f(obj))'

Do you think it will suffice to just create a ticket to investigate its use and replace with the new outputter????

File: src/foam/u2/view/UnstyledTableView.js (hunk 12 above)
Comment (Nauna) [reply src/foam/u2/view/UnstyledTableView.js]: interesting

File: src/foam/lib/csv/Outputter.java (hunk 10 above)
Comment (mcarcaso) [reply src/foam/lib/csv/Outputter.java]: Yeah the CSVSink is basically an Outputter. Let's do this:
- Rename CSVSink to CSVOutputter
- Rename put to outputFObject (so it's a similar interface to json Outputter)
- Rename reset to flush
- Implement toString which returns csv in js and and sb.toString in java

Then implement a CSVSink which creates an instance of an outputter and on put, calls outputFObject and on eof, does a csv = outputter.toString()

That sound ok?

Comment (mcarcaso) [reply src/foam/lib/csv/Outputter.java]: So CSVSink turns to something like this (note that there's no javaCode in here which you'll need to fill in):
```
foam.CLASS({
  package: 'foam.dao',
  name: 'CSVSink',
  extends: 'foam.dao.AbstractSink',

  documentation: 'Sink runs the csv outputter, and contains the resulting string in this.csv',

  properties: [
    {
      class: 'String',
      name: 'csv',
      view: 'foam.u2.tag.TextArea'
    },
    {
      class: 'Class',
      name: 'of',
      visibility: 'HIDDEN'
    },
    {
      name: 'props',
      visibility: 'HIDDEN'
    },
    {
      name: 'outputter',
      transient: true,
      factory: function(props) {
        return this.CSVOutputter.create({
          of: this.of,
          props: this.props
        })
      },
    }
  ],

  methods: [
    function put(obj) {
      this.outputter.outputFObject(obj);
    },

    function eof() {
      this.csv = this.outputter.toString();
    },

    function reset() {
      this.outputter.flush();
      this.csv = '';
    }
  ]
});
```

File: src/foam/dao/CSVSink.js [hunk 13]
```diff
@@ -23,87 +24,58 @@ foam.CLASS({
...
+      of: 'foam.lib.csv.CSVOutputter',
+      transient: true,
+      factory: function() {
+        return foam.lib.csv.CSVOutputter.create({
+          of: this.of,
+          props: this.props
+        });
+      },
+      javaFactory: `
+        return new foam.lib.csv.CSVOutputter.Builder(getX()).setOf(getOf()).setProps(getProps()).build();
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Style suggestion: If you're setting more than one field, break into multiple lines:
```
return new foam.lib.csv.CSVOutputter.Builder(getX())
  .setOf(getOf())
  .setProps(getProps())
  .build();
```
Is much easier to read and maintain.

File: src/foam/lib/csv/CSVOutputter.js [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+    'java.lang.String',
+    'java.util.Date'
+  ],
+
+  properties: [
+    {
+      class: 'String',
+      name: 'csv',
+      view: 'foam.u2.tag.TextArea',
+      documentation: 'Used for js version of this outputter.'
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: If this is for js only then give it `flags: ['js']` and it won't be generated for java.

File: src/foam/lib/csv/CSVOutputter.js [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+      name: 'of',
+      visibility: 'HIDDEN'
+    },
+    {
+      class: 'StringArray',
+      name: 'props',
+      factory: function() {
+        return this.of.getAxiomByName('tableColumns').columns;
+      },
+      visibility: 'HIDDEN'
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: I'm not sure you should hide any of these properties. If I'm viewing the CSVOutputter from a DetailView then it's probably because I'm debugging and I'll want to see a lot of these fields.

File: src/foam/lib/csv/CSVOutputter.js [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+        });
+        this.newLine_();
+        this.isFirstRow = false;
+      },
+      javaCode: `
+        Object propObj;
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        for (String propName: tableColumnNames) {
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: you can just use `getProps()` here and not declare a tableColumnNames above.

File: src/foam/lib/csv/CSVOutputter.js [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+      },
+      javaCode: 'return getSb().toString();'
+    },
+    {
+      name: 'headerOutput',
+      args: [
+        { type: 'FObject', name: 'obj' }
+      ],
+      code: function() {
+        var element;
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Why is this declared up here? It's only used in the loop below.

File: src/foam/lib/csv/CSVOutputter.js [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        this.props.forEach((name) => {
+          element = this.of.getAxiomByName(name);
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Rename element to axiom or prop.

File: src/foam/lib/csv/CSVOutputter.js [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+    },
+    {
+      name: 'toString',
+      code: function() {
+        return this.csv;
+      },
+      javaCode: 'return getSb().toString();'
+    },
+    {
+      name: 'headerOutput',
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: rename this to outputHeader

File: src/foam/lib/csv/CSVOutputter.js [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+        setIsFirstRow(false);
+      `
+    },
+    {
+      name: 'outputFObject',
+      args: [
+        { type: 'FObject', name: 'obj' }
+      ],
+      code: function(obj) {
+        var element;
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Don't declare up here

File: src/foam/lib/csv/CSVOutputter.js [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+
+        if ( ! this.of ) this.of = obj.cls_;
+
+        if ( this.isFirstRow ) this.headerOutput();
+
+        this.props.forEach((name) => {
+          element = this.of.getAxiomByName(name);
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: rename to axiom/prop

File: src/foam/lib/csv/CSVOutputter.js [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+          element.toCSV(x, obj, this, element);
+        });
+
+        this.newLine_();
+      },
+      javaCode: `
+        Object propObj;
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Don't declare any of these variables up here

File: src/foam/lib/csv/CSVOutputter.js [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        if ( ! isPropertySet("of") || getOf() == null ) setOf(obj.getClassInfo());
+
+        if ( getIsFirstRow() ) headerOutput(obj);
+
+        for (String propName : tableColumnNames) {
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: just use getProps()

File: src/foam/lib/csv/CSVOutputter.js [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+        });
+
+        this.newLine_();
+      },
+      javaCode: `
+        Object propObj;
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        if ( ! isPropertySet("of") || getOf() == null ) setOf(obj.getClassInfo());
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Why do you need to `isPropertySet` check?

File: src/foam/core/PropertyInfo.java [hunk 25]
```diff
@@ -43,7 +44,8 @@
   public Parser queryParser();
   public Parser csvParser();
   public void toJSON(foam.lib.json.Outputter outputter, Object value);
-  public void toCSV(foam.lib.csv.Outputter outputter, Object value);
+  public void toCSV(X x, Object obj, foam.lib.csv.CSVOutputter outputter, Object value);
```
Comment (mcarcaso) [inline src/foam/core/PropertyInfo.java]: I think we can do this:
`public void toCSV(X x, FObject obj, foam.lib.csv.CSVOutputter outputter);`
And the toCSV method can figure out the value by doing an `Object value = f(obj);`
`PropertyInfos` implement `Expr` which has an `f` method on it that takes an object and gives you an object. In the case of a PropertyInfo, it takes an FObject and gives you the value of that property on the object.

File: src/foam/core/PropertyInfo.java [hunk 26]
```diff
@@ -43,7 +44,8 @@
...
+  public void toCSVLabel(foam.lib.csv.CSVOutputter outputter, Object value);
```
Comment (mcarcaso) [inline src/foam/core/PropertyInfo.java]: I don't think you need the `value` in this method at all do you?

File: src/foam/lib/csv/CSVOutputter.js [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+        this.newLine_();
+        this.isFirstRow = false;
+      },
+      javaCode: `
+        Object propObj;
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        for (String propName: tableColumnNames) {
+          propObj = obj.getProperty(propName);
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Since this code block is basically identical to the JS version, try to use the same variable names for the same object if you can. columnProp would be renamed to axiom or prop depending on what you renamed it to above.
propObj shouldn't be necessary.

File: src/foam/lib/csv/CSVOutputter.js [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+        Object propObj;
+        PropertyInfo columnProp;
+        String[] tableColumnNames = getProps();
+
+        if ( ! isPropertySet("of") || getOf() == null ) setOf(obj.getClassInfo());
+
+        if ( getIsFirstRow() ) headerOutput(obj);
+
+        for (String propName : tableColumnNames) {
+          propObj = obj.getProperty(propName);
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: Similar comment here as above. Ditch propObj and rename columnProp to axiom/prop.

File: src/foam/lib/csv/Outputter.java (hunk 10 above)
Comment (mcarcaso) [inline src/foam/lib/csv/Outputter.java]: This change can change the behaviour of anything using this Outputter. Please make it a priority to remove this Outputter ASAP so nobody tries to fix it if/when something breaks. Anything using this outputter needs to be updated to use the new outputter.

File: src/foam/nanos/export/CSVDriver.js [hunk 29]
```diff
@@ -10,30 +10,37 @@ foam.CLASS({
...
+      factory: function() {
+        return this.CSVOutputter.create();
+      }
     }
   ],
 
   methods: [
     function exportFObject(X, obj) {
-      return this.outputter.toCSV(obj);
+      // if obj coming in is not from a model with tableColumns,
```
Comment (mcarcaso) [inline src/foam/nanos/export/CSVDriver.js]: I don't think this comment adds much value.

File: src/foam/u2/view/UnstyledTableView.js [hunk 30]
```diff
@@ -53,21 +54,28 @@ foam.CLASS({
     },
     {
       name: 'columns_',
+      documentation: `Note: filteredTableColumns was a property set in DAOController
```
Comment (mcarcaso) [inline src/foam/u2/view/UnstyledTableView.js:57]: This shouldn't be intertwined within this expression like this. Let's use a reaction instead. Add this to the model:
```
reactions: [
  ['', 'propertyChange.columns_', 'updateFilteredTableColumns']
],
listeners: [
  {
    name: 'updateFilteredTableColumns',
    code: function() {
      this.filteredTableColumns = this.columns_.map(a => a.name);
    }
  }
]
```

File: src/foam/core/PropertyInfo.java (hunk 26 above)
Comment (Nauna) [reply src/foam/core/PropertyInfo.java]: No we don't. Thx for the reminder

File: src/foam/u2/view/UnstyledTableView.js (hunk 30 above)
Comment (Nauna) [reply src/foam/u2/view/UnstyledTableView.js:57]: considering this should have been done weeks ago. Are you ok with me making a ticket and adjusting this after.?

Comment (mcarcaso) [reply src/foam/u2/view/UnstyledTableView.js:57]: Why not just do it now? I pretty much gave you the code you need to use above. It might even be less work to just use that than to create a ticket and describe what needs to be done.

**Comment by Nauna** [comment]
Test Cases:
- Dig: calls CSVSink and CSVOutputter
(To Test) ```1)  http://localhost:8080/service/dig?dao=accountDAO&cmd=select&format=csv```
```2)  http://localhost:8080/service/dig?dao=invoiceDAO&cmd=select&format=csv```
(Note on Test 2) accountDAO call will check defaults but testing invoiceDAO tests nested FObject properties.
- EFTTest : uses CSVUtil
- Export from localhost -> admin -> services -> transactionDAO : uses ExportDriver/ExportModal
- Ablii -> payable -> export as CSV: uses CSVDriver

File: src/foam/dao/CSVSink.js [hunk 31]
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
       visibility: 'HIDDEN'
     },
     {
+      class: 'StringArray',
       name: 'props',
-      expression: function(of) {
-        return of.getAxiomsByClass(foam.core.Property)
-          .filter( (p) => ! p.networkTransient );
+      factory: function(of) {
+        if ( ! of ) return;
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Maybe return `[]`?

File: src/foam/dao/CSVSink.js [hunk 32]
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
...
-      visibility: 'HIDDEN'
-    },
-    {
-      class: 'Boolean',
-      name: 'isHeadersOutput',
-      visibility: 'HIDDEN'
+      javaFactory: `
+        if ( ! isPropertySet("of") ) {
+          propsIsSet_ = false;
+          return null;
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Same here. Rather than marking it as unset just return an empty array? I'm hesitant to mess with the propIsSet_ cuz that's managed by the model. Also, I could be wrong but I think this will get clobbered right after the factory runs and set to true.

File: src/foam/dao/CSVSink.js [hunk 33]
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
...
     {
-      class: 'Boolean',
-      name: 'isNewLine',
-      value: true,
-      visibility: 'HIDDEN'
+      name: 'outputter',
+      class: 'FObjectProperty',
+      of: 'foam.lib.csv.CSVOutputter',
+      transient: true,
+      factory: function(of, props) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: These arguments won't actually be getting passed in since this is a factory. Ditch these from here and use this.of and this.props below

File: src/foam/dao/CSVSink.js [hunk 34]
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
...
+        var csvOutputter = foam.lib.csv.CSVOutputter.create();
+        if ( of ) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Is there any harm in removing these checks and just creating the CSVOutputter like:
```
this.CSVOutputter.create({
  of: this.of,
  props: this.props
})
```
?

File: src/foam/dao/CSVSink.js [hunk 35]
```diff
@@ -23,87 +32,88 @@ foam.CLASS({
...
+          csvOutputter.of = of;
+        }
+        if ( props ) {
+          csvOutputter.props = props;
+        }
+        return csvOutputter;
+      },
+      javaFactory: `
+      foam.lib.csv.CSVOutputter csvOutputter = new foam.lib.csv.CSVOutputter.Builder(getX()).build();
+      if ( isPropertySet("of") ) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Do this without the conditional if you can too.
```
foam.lib.csv.CSVOutputter csvOutputter = new foam.lib.csv.CSVOutputter.Builder(getX())
  .setOf(getOf())
  .setProps(getProps())
  .build();
```

File: src/foam/dao/CSVSink.js [hunk 36]
```diff
@@ -120,15 +130,15 @@ foam.CLASS({
     {
       name: 'toCSV',
       class: 'Function',
-      value: function(obj, outputter, prop) {
-        outputter.output(obj ? obj[prop.name] : null);
+      value: function(obj, outputter) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: x as first arg?

File: src/foam/dao/CSVSink.js [hunk 37]
```diff
@@ -120,15 +130,15 @@ foam.CLASS({
...
+        outputter.outputValue(obj ? obj[this.name] : null);
       }
     },
     {
       name: 'toCSVLabel',
       class: 'Function',
-      value: function(outputter, prop) {
-        outputter.output(prop.name);
+      value: function(x, outputter, obj) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: I don't think you need/want obj here. The code would be:
`outputter.outputValue(this.name);`
(where `this` is the foam.core.Property)

File: src/foam/dao/CSVSink.js [hunk 38]
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
...
+      value: function(x, obj, outputter) {
+        if ( ! this.of ) {
+          outputter.outputValue(obj ? obj[this.name] : null);
           return;
         }
-        prop.of.getAxiomsByClass(foam.core.Property)
+        this.of.getAxiomsByClass(foam.core.Property)
           .forEach((axiom) => {
-            axiom.toCSV(obj ? obj[prop.name] : null, outputter, axiom);
+            axiom.toCSV(x, obj ? obj[this.name] : null, outputter, axiom);
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: I think you want:
`axiom.toCSV.call(axiom, x, obj && obj[this.name], outputter)`

File: src/foam/dao/CSVSink.js [hunk 39]
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
...
           });
       }
     },
+    {
+      name: 'javaToCSV',
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js:157]: Add these refinements in java/refinements.js

File: src/foam/dao/CSVSink.js [hunk 40]
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
...
+      class: 'String',
+      value: `
+        if ( get(obj) == null && ! (get(obj) instanceof foam.core.FObject) ) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: This check isn't right. You should be doing the same check as js. If there's no `of` then you just output the object as is.

File: src/foam/dao/CSVSink.js [hunk 41]
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
...
+          outputter.outputValue(getName());
+          return;
+        }
+
+        java.util.List<foam.core.PropertyInfo> nestPropList = ((foam.core.FObject)get(obj)).getClassInfo().getAxiomsByClass(foam.core.PropertyInfo.class);
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Here too, you want to base it off the `of` of the FObjectProperty. Using the ClassInfo of the property being stored can lead to bugs if it's a subclass of whatever the `of` is.

File: src/foam/dao/CSVSink.js [hunk 42]
```diff
@@ -147,36 +157,70 @@ foam.CLASS({
...
+        for ( foam.core.PropertyInfo axiom : nestPropList ) {
+          axiom.toCSV(x, get(obj), outputter);
+        }
+      `
+    },
     {
       name: 'toCSVLabel',
       class: 'Function',
-      value: function(outputter, prop) {
+      value: function(x, outputter, obj) {
```
Comment (mcarcaso) [inline src/foam/dao/CSVSink.js]: Shouldn't need `obj` here. Base decisions on this.of

File: src/foam/java/PropertyInfo.js [hunk 43]
```diff
@@ -285,6 +287,28 @@ foam.CLASS({
...
+            body: this.toCSV
+          },
+          {
+            name: 'toCSVLabel',
+            visibility: 'public',
+            type: 'void',
+            args: [
+              { name: 'x',          type: 'foam.core.X' },
+              { name: 'outputter',  type: 'foam.lib.csv.CSVOutputterInterface' },
+              { name: 'obj',        type: 'Object' }
```
Comment (mcarcaso) [inline src/foam/java/PropertyInfo.js]: You shouldn't need an obj here.

File: src/foam/lib/csv/CSVOutputter.js [hunk 44]
```diff
@@ -0,0 +1,235 @@
...
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.lib.csv',
+  name: 'CSVOutputter',
+
+  implements: [
+    'foam.lib.csv.CSVOutputterInterface'
```
Comment (mcarcaso) [inline src/foam/lib/csv/CSVOutputter.js]: I think the more common/consistent way to do this would be to have the interface called CSVOutputter and this model called CSVOutputterImpl

File: src/foam/lib/csv/FObjectCSVOutputterDecorator.js [hunk 45]
```diff
@@ -0,0 +1,41 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+foam.CLASS({
+  package: 'foam.lib.csv',
+  name: 'FObjectCSVOutputterDecorator',
```
Comment (mcarcaso) [inline src/foam/lib/csv/FObjectCSVOutputterDecorator.js]: Create a ProxyCSVOutputter and make this extend that. This way, we can add methods to the interface and nothing else should need to get outputted.

Comment (mcarcaso) [reply src/foam/lib/csv/FObjectCSVOutputterDecorator.js]: Also rename this to PrefixedValueCSVOutputter ?

File: src/foam/lib/csv/FObjectCSVOutputterDecorator.js [hunk 46]
```diff
@@ -0,0 +1,41 @@
...
+  ],
+
+  properties: [
+    {
+      name: 'outputter',
+      class: 'FObjectProperty',
+      of: 'foam.lib.csv.CSVOutputterInterface'
+    },
+    {
+      name: 'preLabelString',
```
Comment (mcarcaso) [inline src/foam/lib/csv/FObjectCSVOutputterDecorator.js]: Just call this `prefix`

File: src/foam/lib/csv/Outputter.java [hunk 47]
```diff
@@ -1,235 +0,0 @@
-/**
- * @license
- * Copyright 2017 The FOAM Authors. All Rights Reserved.
```
Comment (mcarcaso) [inline src/foam/lib/csv/Outputter.java:3]: Woohoo!

**Review by mcarcaso — APPROVED** [review]
More featured and less code. Beautiful :)
